utils/bignum/complex.go

```diff
@@ -0,0 +1,226 @@
+package bignum
+
+import (
+	"fmt"
+	"math/big"
+)
+
+// Complex is a type for arbitrary precision complex number
+type Complex [2]*big.Float
+
+// NewComplex creates a new arbitrary precision complex number
+func NewComplex() (c *Complex) {
+	return &Complex{
+		new(big.Float),
+		new(big.Float),
+	}
+}
+
+// ToComplex takes a complex128, float64, int, int64, uint, uint64, *big.Int, *big.Float or *Complex and returns a *Complex set to the given precision.
+func ToComplex(value interface{}, prec uint) (cmplx *Complex) {
+
+	cmplx = new(Complex)
+
+	switch value := value.(type) {
+	case complex128:
+		cmplx[0] = new(big.Float).SetPrec(prec).SetFloat64(real(value))
+		cmplx[1] = new(big.Float).SetPrec(prec).SetFloat64(imag(value))
+	case float64:
+		cmplx[0] = new(big.Float).SetPrec(prec).SetFloat64(value)
+		cmplx[1] = new(big.Float).SetPrec(prec)
+	case int:
+		cmplx[0] = new(big.Float).SetPrec(prec).SetInt64(int64(value))
+		cmplx[1] = new(big.Float).SetPrec(prec)
+	case int64:
+		cmplx[0] = new(big.Float).SetPrec(prec).SetInt64(value)
+		cmplx[1] = new(big.Float).SetPrec(prec)
+	case uint64:
+		return ToComplex(new(big.Int).SetUint64(value), prec)
+	case *big.Float:
+		cmplx[0] = new(big.Float).SetPrec(prec).Set(value)
+		cmplx[1] = new(big.Float).SetPrec(prec)
+	case *big.Int:
+		cmplx[0] = new(big.Float).SetPrec(prec).SetInt(value)
+		cmplx[1] = new(big.Float).SetPrec(prec)
+	case *Complex:
+		cmplx[0] = new(big.Float).SetPrec(prec).Set(value[0])
+		cmplx[1] = new(big.Float).SetPrec(prec).Set(value[1])
+	default:
+		panic(fmt.Errorf("invalid value.(type): must be int, int64, uint64, float64, complex128, *big.Int, *big.Float or *Complex but is %T", value))
+	}
+
+	return
+}
+
+// IsInt returns true if both the real and imaginary part are integers.
+func (c *Complex) IsInt() bool {
+	return c[0].IsInt() && c[1].IsInt()
+}
+
+func (c *Complex) IsReal() bool {
+	return c[1].Cmp(new(big.Float)) == 0
+}
+
+func (c *Complex) SetComplex128(x complex128) *Complex {
+	c[0].SetFloat64(real(x))
+	c[1].SetFloat64(imag(x))
+	return c
+}
+
+// Set sets a arbitrary precision complex number
+func (c *Complex) Set(a *Complex) *Complex {
+	c[0].Set(a[0])
+	c[1].Set(a[1])
+	return c
+}
+
+func (c *Complex) Prec() uint {
+	return c[0].Prec()
```

Comment:
What if c[1].Prec() < c[0].Prec()?
Or is it impossible?

Reply:
Done